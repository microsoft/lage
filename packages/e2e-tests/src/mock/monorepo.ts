import os from "os";
import fs from "fs";
import path from "path";
import execa from "execa";

export class Monorepo {
  static tmpdir = os.tmpdir();

  root: string;
  nodeModulesPath: string;
  lagePath: string;

  constructor(private name: string) {
    this.root = fs.mkdtempSync(path.join(Monorepo.tmpdir, `lage-monorepo-${name}-`));
    this.nodeModulesPath = path.join(this.root, "node_modules");
    this.lagePath = path.join(this.nodeModulesPath, "@lage-run", "cli");
  }

  init() {
    const options = { cwd: this.root };
    execa.sync("git", ["init"], options);
    execa.sync("git", ["config", "user.email", "you@example.com"], options);
    execa.sync("git", ["config", "user.name", "test user"], options);
    execa.sync("git", ["config", "commit.gpgsign", "false"], options);
    this.generateRepoFiles();
  }

  install() {
    if (!fs.existsSync(this.nodeModulesPath)) {
      fs.mkdirSync(this.nodeModulesPath, { recursive: true });
    }

    // pretends to perform a npm install of lage
    if (!fs.existsSync(this.lagePath)) {
      fs.mkdirSync(path.dirname(this.lagePath), { recursive: true });
      fs.symlinkSync(path.join(__dirname, "..", "..", "..", "cli"), this.lagePath, "junction");
    }
  }

  /**
   * Simulates a "yarn" call by linking internal packages and generates a yarn.lock file
   */
  linkPackages() {
    const pkgs = fs.readdirSync(path.join(this.root, "packages"));

    if (!fs.existsSync(this.nodeModulesPath)) {
      fs.mkdirSync(this.nodeModulesPath, { recursive: true });
    }

    let yarnYaml = `# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.\n# yarn lockfile v1\n`;

    for (const pkg of pkgs) {
      fs.symlinkSync(path.join(this.root, "packages", pkg), path.join(this.nodeModulesPath, pkg), "junction");

      const pkgJson = JSON.parse(fs.readFileSync(path.join(this.root, "packages", pkg, "package.json"), "utf-8"));
      const deps = pkgJson.dependencies;

      yarnYaml += `"${pkg}@^${pkgJson.version}":\n  version "${pkgJson.version}"\n`;

      if (deps) {
        yarnYaml += `  dependencies:`;
        for (const dep of Object.keys(deps)) {
          yarnYaml += `    "${dep}" "0.1.0"`;
        }
      }
    }

    this.commitFiles({ "yarn.lock": yarnYaml });
  }

  generateRepoFiles() {
    const lagePathBin = `${this.lagePath}/lib/cli.js`;

    this.commitFiles({
      "package.json": {
        name: this.name,
        version: "0.1.0",
        private: true,
        workspaces: ["packages/*"],
        scripts: {
          bundle: `node "${lagePathBin}" bundle --reporter json --log-level silly`,
          transpile: `node "${lagePathBin}" transpile --reporter json --log-level silly`,
          build: `node "${lagePathBin}" build --reporter json --log-level silly`,
          writeInfo: `node "${lagePathBin}" info`,
          test: `node "${lagePathBin}" test --reporter json --log-level silly`,
          lint: `node "${lagePathBin}" lint --reporter json --log-level silly`,
          clear: `node "${lagePathBin}" cache --clear --reporter json --log-level silly`,
        },
        devDependencies: {
          lage: path.resolve(__dirname, "..", ".."),
        },
      },
      "lage.config.js": `module.exports = {
        pipeline: {
          build: ['^build'],
          test: ['build'],
          lint: []
        }
      };`,
      ".gitignore": "node_modules",
    });
  }

  setLageConfig(contents: string) {
    this.commitFiles({
      "lage.config.js": contents,
    });
  }

  addPackage(name: string, internalDeps: string[] = [], scripts?: { [script: string]: string }) {
    return this.commitFiles({
      [`packages/${name}/build.js`]: `console.log('building ${name}');`,
      [`packages/${name}/test.js`]: `console.log('building ${name}');`,
      [`packages/${name}/lint.js`]: `console.log('linting ${name}');`,
      [`packages/${name}/package.json`]: {
        name,
        version: "0.1.0",
        scripts: scripts || {
          build: "node ./build.js",
          test: "node ./test.js",
          lint: "node ./lint.js",
        },
        dependencies: {
          ...(internalDeps &&
            internalDeps.reduce((deps, dep) => {
              return { ...deps, [dep]: "*" };
            }, {})),
        },
      },
    });
  }

  clone(origin: string) {
    return execa.sync("git", ["clone", origin], { cwd: this.root });
  }

  push(origin: string, branch: string) {
    return execa.sync("git", ["push", origin, branch], { cwd: this.root });
  }

  commitFiles(files: { [name: string]: string | Record<string, unknown> }, options: { executable?: boolean } = {}) {
    for (const [file, contents] of Object.entries(files)) {
      let out = "";
      if (typeof contents !== "string") {
        out = JSON.stringify(contents, null, 2);
      } else {
        out = contents;
      }

      const fullPath = path.join(this.root, file);

      if (!fs.existsSync(path.dirname(fullPath))) {
        fs.mkdirSync(path.dirname(fullPath), { recursive: true });
      }

      fs.writeFileSync(fullPath, out);

      if (options.executable) {
        fs.chmodSync(path.join(this.root, file), fs.constants.S_IXUSR | fs.constants.S_IRUSR | fs.constants.S_IROTH);
      }
    }

    execa.sync("git", ["add", ...Object.keys(files)], {
      cwd: this.root,
    });

    execa.sync("git", ["commit", "-m", "commit files"], { cwd: this.root });
  }

  run(command: string, args?: string[], silent?: boolean) {
    return execa.sync("yarn", [...(silent === true ? ["--silent"] : []), command, ...(args || [])], {
      cwd: this.root,
    });
  }

  cleanup() {
    fs.rmSync(this.root, { recursive: true });
  }
}
