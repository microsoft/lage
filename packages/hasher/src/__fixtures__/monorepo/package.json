{
  "name": "monorepo",
  "license": "MIT",
  "version": "0.1.0",
  "workspaces": {
    "packages": [
      "packages/*"
    ]
  }
}
