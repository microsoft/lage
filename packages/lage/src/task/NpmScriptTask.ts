import { TaskLogger } from "../logger/TaskLogger";
import { ChildProcess } from "child_process";
import { PackageInfo } from "workspace-tools";
import { findNpmClient } from "../workspace/findNpmClient";
import { spawn } from "child_process";
import path from "path";
import { getNpmCommand } from "./getNpmCommand";
import { Config } from "../types/Config";
import { LogLevel } from "../logger/LogLevel";

export class NpmScriptTask {
  static npmCmd: string = "";
  static activeProcesses = new Set<ChildProcess>();
  static gracefulKillTimeout = 2500;

  npmArgs: string[] = [];
  startTime: [number, number] = [0, 0];
  duration: [number, number] = [0, 0];

  static killAllActiveProcesses() {
    // first, send SIGTERM everywhere
    for (const cp of NpmScriptTask.activeProcesses) {
      cp.kill("SIGTERM");
    }

    // wait for "gracefulKillTimeout" to make sure everything is terminated via SIGKILL
    setTimeout(() => {
      for (const cp of NpmScriptTask.activeProcesses) {
        if (!cp.killed) {
          cp.kill("SIGKILL");
        }
      }
    }, NpmScriptTask.gracefulKillTimeout);
  }

  constructor(public task: string, public info: PackageInfo, private config: Config, private logger: TaskLogger) {
    NpmScriptTask.npmCmd = NpmScriptTask.npmCmd || findNpmClient(config.npmClient);
    this.npmArgs = getNpmCommand(config.node, config.args, task);
  }

  run() {
    const { info, logger, npmArgs } = this;
    const { npmCmd } = NpmScriptTask;
    return new Promise<void>((resolve, reject) => {
      logger.verbose(`Running ${[npmCmd, ...npmArgs].join(" ")}`);

      const cp = spawn(npmCmd, npmArgs, {
        cwd: path.dirname(info.packageJsonPath),
        stdio: "pipe",
        env: {
          ...process.env,
          ...(process.stdout.isTTY && this.config.reporter !== "json" && { FORCE_COLOR: "1" }),
          LAGE_PACKAGE_NAME: info.name,
        },
      });

      NpmScriptTask.activeProcesses.add(cp);

      this.logger.stream(LogLevel.verbose, cp.stdout);
      this.logger.stream(LogLevel.verbose, cp.stderr);
      cp.on("exit", handleChildProcessExit);

      function handleChildProcessExit(code: number) {
        if (code === 0) {
          NpmScriptTask.activeProcesses.delete(cp);
          return resolve();
        }

        cp.stdout.destroy();
        cp.stdin.destroy();
        reject();
      }
    });
  }
}
