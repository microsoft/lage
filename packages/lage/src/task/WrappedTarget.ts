import { TaskLogger } from "../logger/TaskLogger";
import { ChildProcess } from "child_process";
import { controller } from "./abortSignal";
import { cacheHash, cacheFetch, cachePut } from "../cache/backfill";
import { RunContext } from "../types/RunContext";
import { hrToSeconds } from "../logger/reporters/formatDuration";
import { Config } from "../types/Config";
import { getPackageAndTask } from "./taskId";
import { CacheOptions } from "../types/CacheOptions";
import { TargetStatus } from "../types/TargetStatus";
import { LoggableTarget, PipelineTarget } from "../types/PipelineDefinition";

export class WrappedTarget implements LoggableTarget {
  static npmCmd: string = "";
  static activeProcesses = new Set<ChildProcess>();
  static gracefulKillTimeout = 2500;

  npmArgs: string[] = [];
  startTime: [number, number] = [0, 0];
  duration: [number, number] = [0, 0];
  status: TargetStatus;
  logger: TaskLogger;
  cacheOptions: CacheOptions;

  constructor(public target: PipelineTarget, private root: string, private config: Config, private context: RunContext) {
    this.status = "pending";
    this.logger = new TaskLogger(target.packageName || "[GLOBAL]", target.packageName ? target.task : target.id);

    const outputGlob = target.outputGlob || config.cacheOptions.outputGlob;
    this.cacheOptions = {
      ...config.cacheOptions,
      ...(outputGlob && { outputGlob }),
    };

    this.context.targets.set(target.id, this);
  }

  onStart() {
    this.status = "started";
    this.startTime = process.hrtime();
    this.logger.info("started", { status: "started" });
  }

  onComplete() {
    this.status = "completed";
    this.duration = process.hrtime(this.startTime);
    this.logger.info("completed", {
      status: "completed",
      duration: hrToSeconds(this.duration),
    });
  }

  onFail() {
    this.status = "failed";
    this.duration = process.hrtime(this.startTime);
    this.logger.info("failed", {
      status: "failed",
      duration: hrToSeconds(this.duration),
    });
  }

  onSkipped(hash: string | null) {
    this.status = "skipped";
    this.duration = process.hrtime(this.startTime);
    this.logger.info(`skipped`, {
      status: "skipped",
      duration: hrToSeconds(this.duration),
      hash,
    });
  }

  async getCache() {
    let hash: string | null = null;
    let cacheHit = false;

    const { target, root, config, cacheOptions } = this;

    if (config.cache && target.cache) {
      hash = await cacheHash(target.id, target.cwd, root, cacheOptions, config.args);

      if (hash && !config.resetCache) {
        cacheHit = await cacheFetch(hash, target.id, target.cwd, cacheOptions);
      }
    }

    return { hash, cacheHit };
  }

  async saveCache(hash: string | null) {
    const { logger, target, cacheOptions } = this;
    logger.verbose(`hash put ${hash}`);
    await cachePut(hash, target.cwd, cacheOptions);
  }

  async run() {
    const { target, context, config, logger } = this;

    try {
      const { hash, cacheHit } = await this.getCache();

      const cacheEnabled = target.cache && config.cache && hash;

      this.onStart();

      // skip if cache hit!
      if (cacheHit) {
        // save the cache anyway - this will cause remote cache to be saved locally
        if (cacheEnabled) {
          await this.saveCache(hash);
        }

        this.onSkipped(hash);
        return true;
      }

      if (cacheEnabled) {
        logger.verbose(`hash: ${hash}, cache hit? ${cacheHit}`);
      }

      // Wraps with profiler as well as task args
      await context.profiler.run(() => {
        if (!target.run) {
          return Promise.resolve();
        }

        let result: Promise<unknown> | void;

        if (target.packageName) {
          result = target.run({
            packageName: target.packageName,
            config: this.config,
            cwd: target.cwd,
            options: target.options,
            taskName: getPackageAndTask(target.id).task,
            logger,
          });
        } else {
          result = target.run({
            config: this.config,
            cwd: target.cwd,
            options: target.options,
            logger,
          });
        }

        if (!result || typeof result["then"] !== "function") {
          return Promise.resolve(result);
        }

        return result as Promise<unknown>;
      }, target.id);

      if (cacheEnabled) {
        await this.saveCache(hash);
      }

      this.onComplete();
    } catch (e) {
      context.measures.failedTargets = context.measures.failedTargets || [];
      context.measures.failedTargets.push(target.id);

      this.onFail();

      if (!config.safeExit && !config.continue) {
        controller.abort();
      }

      throw e;
    }

    return true;
  }
}
