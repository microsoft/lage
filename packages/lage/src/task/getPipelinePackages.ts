import { Config } from "../types/Config";
import { filterPackages } from "./filterPackages";
import { Workspace } from "../types/Workspace";
import { getScopedPackages, getChangedPackages, getBranchChanges } from "workspace-tools";
import * as fg from "fast-glob";
import { logger } from "../logger";

export function getPipelinePackages(workspace: Workspace, config: Config) {
  const { scope, since, repoWideChanges, includeDependencies } = config;

  // If scoped is defined, get scoped packages
  const hasScopes = Array.isArray(scope) && scope.length > 0;
  let scopedPackages: string[] | undefined = undefined;
  if (hasScopes) {
    scopedPackages = getScopedPackages(scope!, workspace.allPackages);
  }

  const hasSince = typeof since !== "undefined";
  let changedPackages: string[] | undefined = undefined;

  // Be specific with the changed packages only if no repo-wide changes occurred
  if (hasSince && !hasRepoChanged(since, workspace.root, repoWideChanges)) {
    try {
      changedPackages = getChangedPackages(workspace.root, since, config.ignore);
    } catch (e) {
      logger.warn(`An error in the git command has caused this scope run to include every package\n${e}`);
      // if getChangedPackages throws, we will assume all have changed (using changedPackage = undefined)
    }
  }

  return filterPackages({
    allPackages: workspace.allPackages,
    deps: config.deps,
    scopedPackages,
    changedPackages,
    includeDependencies,
  });
}

function hasRepoChanged(since: string, root: string, environmentGlob: string[]) {
  try {
    const changedFiles = getBranchChanges(since, root);
    const envFiles = fg.sync(environmentGlob, { cwd: root });
    let repoWideChanged = false;

    if (changedFiles) {
      for (const change of changedFiles) {
        if (envFiles.includes(change)) {
          repoWideChanged = true;
          break;
        }
      }
    }

    return repoWideChanged;
  } catch (e) {
    // if this fails, let's assume repo has changed
    logger.warn(`An error in the git command has caused this to consider the repo has changed\n${e}`);
    return true;
  }
}
