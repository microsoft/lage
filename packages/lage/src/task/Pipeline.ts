import { Config } from "../types/Config";
import { generateTopologicGraph } from "../workspace/generateTopologicalGraph";
import { NpmScriptTask } from "./NpmScriptTask";
import { PackageInfo, PackageInfos } from "workspace-tools";
import { RunContext } from "../types/RunContext";
import { PipelineTarget, TargetConfig, TargetConfigFactory, TaskArgs } from "../types/PipelineDefinition";
import { TopologicalGraph } from "../types/TopologicalGraph";
import { Workspace } from "../types/Workspace";
import pGraph, { PGraphNodeMap } from "p-graph";
import path from "path";
import { getPipelinePackages } from "./getPipelinePackages";
import { getPackageAndTask, getTargetId } from "./taskId";
import { WrappedTarget } from "./WrappedTarget";

export const START_TARGET_ID = "__start";

/**
 * Pipeline class represents lage's understanding of the dependency graphs and wraps the promise graph implementations to execute tasks in order
 */
export class Pipeline {
  private cachedTransitiveTaskDependencies: Map<string, "walk-in-progress" | Set<string>> = new Map();

  /** Target represent a unit of work and the configuration of how to run it */
  targets: Map<string, PipelineTarget> = new Map([
    [
      START_TARGET_ID,
      {
        id: START_TARGET_ID,
        cwd: "",
        run: () => {},
        task: START_TARGET_ID,
        hidden: true,
        cache: false,
      },
    ],
  ]);

  /** Target dependencies determine the run order of the targets  */
  dependencies: [string, string][] = [];

  /** Internal cache of the package.json information */
  packageInfos: PackageInfos;

  /** Internal generated cache of the topological package graph */
  graph: TopologicalGraph;

  /** Internal cache of context */
  context: RunContext | undefined;

  constructor(private workspace: Workspace, private config: Config) {
    this.packageInfos = workspace.allPackages;
    this.graph = generateTopologicGraph(workspace);
    this.loadConfig(config);
  }

  /**
   * NPM Tasks are blindly placed in the task dependency graph, but we skip doing work if the package does not contain the specific npm task
   * @param task
   * @param info
   * @returns
   */
  private maybeRunNpmTask(task: string, info: PackageInfo) {
    if (!info.scripts?.[task]) {
      return;
    }

    return (args: TaskArgs) => {
      const npmTask = new NpmScriptTask(task, info, this.config, args.logger);
      return npmTask.run();
    };
  }

  /**
   * Generates a package target during the expansion of the shortcut syntax
   */
  private generatePackageTarget(packageName: string, task: string, deps: string[]): PipelineTarget {
    const info = this.packageInfos[packageName];
    const id = getTargetId(packageName, task);

    return {
      id,
      task,
      cache: this.config.cache,
      outputGlob: this.config.cacheOptions.outputGlob,
      packageName: packageName,
      cwd: path.dirname(this.packageInfos[packageName].packageJsonPath),
      run: this.maybeRunNpmTask(task, info),
      // TODO: do we need to really merge this? Is this desired? (this is the OLD behavior)
      deps: this.targets.has(id) ? [...(this.targets.get(id)!.deps || []), ...deps] : deps,
    };
  }

  /**
   * Expands the shorthand notation to pipeline targets (executable units)
   */
  private expandShorthandTargets(id: string, deps: string[]): PipelineTarget[] {
    // shorthand gets converted to npm tasks
    const { packageName, task } = getPackageAndTask(id);
    const results: PipelineTarget[] = [];
    let packages: string[] = [];

    if (packageName) {
      // specific case in definition (e.g. 'package-name#test': ['build'])
      packages.push(packageName);
    } else {
      // generic case in definition (e.g. 'test': ['build'])
      packages = Object.entries(this.packageInfos).map(([pkg, _info]) => pkg);
    }

    for (const packageWithScript of packages) {
      results.push(this.generatePackageTarget(packageWithScript, task, deps));
    }

    return results;
  }

  /**
   * Given an id & factory, generate targets configurations
   * @param id
   * @param factory
   */
  private generateFactoryTargets(factory: TargetConfigFactory): TargetConfig[] {
    const targets = factory({
      config: this.config,
      cwd: this.workspace.root,
    });

    return Array.isArray(targets) ? targets : [targets];
  }

  /**
   * Converts target configuration to pipeline targets
   * @param id
   * @param target
   */
  private convertToPipelineTarget(id: string, index: number, target: TargetConfig): PipelineTarget[] {
    if (target.type === "global") {
      const targetId = `${id}.${index}`;
      return [
        {
          ...target,
          id: targetId,
          cache: target.cache !== false,
          cwd: this.workspace.root,
          task: id,
          run: () => {},
        },
      ];
    } else if (id.includes("#")) {
      const { packageName: pkg, task } = getPackageAndTask(id);
      return [
        {
          ...target,

          id,
          cache: target.cache !== false,
          task: id,
          cwd: path.dirname(this.packageInfos[pkg!].packageJsonPath),
          packageName: pkg,
          run: this.maybeRunNpmTask(task, this.packageInfos[pkg!]),
        },
      ];
    } else {
      const packages = Object.entries(this.packageInfos);
      return packages.map(([pkg, _info]) => {
        const targetId = getTargetId(pkg, id);
        return {
          ...target,

          id: targetId,
          cache: target.cache !== false,
          task: id,
          cwd: path.dirname(this.packageInfos[pkg].packageJsonPath),
          packageName: pkg,
          run: this.maybeRunNpmTask(id, this.packageInfos[pkg]),
        };
      });
    }
  }

  /**
   * Adds a target definition (takes in shorthand, target config, or a target config factory)
   * @param id
   * @param targetDefinition
   */
  addTargetDefinition(id: string, targetDefinition: string[] | TargetConfig | TargetConfigFactory) {
    // e.g. build: ["^build", "prepare"]
    if (Array.isArray(targetDefinition)) {
      const targets = this.expandShorthandTargets(id, targetDefinition);

      for (const target of targets) {
        this.targets.set(target.id!, target);
      }
    } else {
      // e.g. build: { /* target config */ }
      const targets = typeof targetDefinition === "function" ? this.generateFactoryTargets(targetDefinition) : [targetDefinition];

      targets.forEach((target, index) => {
        const pipelineTargets = this.convertToPipelineTarget(id, index, target);

        for (const pipelineTarget of pipelineTargets) {
          this.targets.set(pipelineTarget.id, pipelineTarget);
        }
      });
    }
  }

  /**
   * Adds all the target dependencies to the graph
   */
  addDependencies() {
    const targets = [...this.targets.values()];

    for (const target of targets) {
      const { deps, packageName, id } = target;

      // Always start with a root node with a special "START_TARGET_ID"
      this.dependencies.push([START_TARGET_ID, id]);

      // Skip any targets that have no "deps" specified
      if (!deps || deps.length === 0) {
        continue;
      }

      /**
       * Now for every deps defined, we need to "interpret" it based on the syntax:
       * - for any deps like package#task, we simply add the singular dependency (source could be a single package or all packages)
       * - for anything that starts with a "^", we add the package-tasks according to the topological package graph
       *    NOTE: in a non-strict mode (TODO), the dependencies can come from transitive task dependencies
       * - for anything that starts with a "^^", we add the package-tasks from the transitive dependencies in the topological
       *    package graph.
       * - for {"pkgA#task": ["dep"]}, we interpret to add "pkgA#dep"
       * - for anything that is a string without a "^", we treat that string as the name of a task, adding all targets that way
       *    NOTE: in a non-strict mode (TODO), the dependencies can come from transitive task dependencies
       *
       * We interpret anything outside of these conditions as invalid
       */
      for (const dep of deps) {
        if (dep.includes("#")) {
          // package and task as deps
          this.dependencies.push([dep, id]);
        } else if (dep.startsWith("^") && packageName) {
          // topo dep -> build: ['^build']
          const [depTask, dependencySet] = dep.startsWith("^^")
            ? [dep.substr(2), [...this.getTransitiveGraphDependencies(packageName)]]
            : [dep.substr(1), this.graph[packageName].dependencies];

          const dependencyIds = targets
            .filter((needle) => {
              const { task, packageName: needlePackageName } = needle;

              return task === depTask && dependencySet.some((depPkg) => depPkg === needlePackageName);
            })
            .map((needle) => needle.id);

          for (const dependencyId of dependencyIds) {
            this.dependencies.push([dependencyId, id]);
          }
        } else if (packageName) {
          // Intra package task dependency - only add the target dependency if it exists in the pipeline targets lists
          if (this.targets.has(getTargetId(packageName, dep))) {
            this.dependencies.push([getTargetId(packageName, dep), target.id]);
          }
        } else if (!dep.startsWith("^")) {
          const dependencyIds = targets.filter((needle) => needle.task === dep).map((needle) => needle.id);

          for (const dependencyId of dependencyIds) {
            this.dependencies.push([dependencyId, id]);
          }
        } else {
          throw new Error(`invalid pipeline config detected: ${target.id}`);
        }
      }
    }
  }

  /**
   * Gets a list of package names that are direct or indirect dependencies of rootPackageName in this.graph,
   * and caches them on the Pipeline.
   * @param packageName the root package to begin walking from
   */
  getTransitiveGraphDependencies(packageName: string): Set<string> {
    const cachedResult = this.cachedTransitiveTaskDependencies.get(packageName);
    if (cachedResult) {
      return cachedResult === "walk-in-progress"
        ? // There is a recursive walk over this set of dependencies in progress.
          // If we hit this case, that means that a dependency of this package depends on it.
          //
          // In this case we return an empty set to omit this package and it's downstream from its
          // dependency
          new Set()
        : // we already computed this for this package, return the cached result.
          cachedResult;
    } else {
      // No cached result. Compute now with a recursive walk

      // mark that we are traversing this package to prevent infinite recursion
      // in cases of circular dependencies
      this.cachedTransitiveTaskDependencies.set(packageName, "walk-in-progress");

      let immediateDependencies = this.graph[packageName]?.dependencies ?? [];

      // build the set of transitive dependencies by recursively walking the
      // immediate dependencies' dependencies.
      let transitiveDepSet = new Set<string>(immediateDependencies);
      for (let immediateDependency of immediateDependencies) {
        for (let transitiveSubDependency of this.getTransitiveGraphDependencies(immediateDependency)) {
          transitiveDepSet.add(transitiveSubDependency);
        }
      }

      // Cache the result and return
      this.cachedTransitiveTaskDependencies.set(packageName, transitiveDepSet);
      return transitiveDepSet;
    }
  }

  generateTargetGraph() {
    const scope = getPipelinePackages(this.workspace, this.config);
    const tasks = this.config.command;

    const targetGraph: [string, string][] = [];

    const queue: string[] = [];

    for (const task of tasks) {
      // package task
      for (const pkg of scope) {
        if (this.targets.has(getTargetId(pkg, task))) {
          queue.push(getTargetId(pkg, task));
          targetGraph.push([START_TARGET_ID, getTargetId(pkg, task)]);
        }
      }

      // if we have globals, send those into the queue
      for (const target of this.targets.values()) {
        if (target.task === task && !target.packageName) {
          queue.push(target.id);
          targetGraph.push([START_TARGET_ID, target.id]);
        }
      }
    }

    const visited = new Set<string>();

    while (queue.length > 0) {
      const id = queue.shift()!;

      if (visited.has(id)) {
        continue;
      }

      visited.add(id);

      const { packageName, task } = getPackageAndTask(id);

      if (!packageName) {
        // global - find all deps in the form of "task.index"
        for (const [from, to] of this.dependencies) {
          if (to.includes(".")) {
            const toTaskName = to.split(".")[0];
            if (toTaskName === task) {
              targetGraph.push([from, to]);

              if (from) {
                queue.push(from);
              }
            }
          }
        }
      } else {
        // package dep
        for (const [from, to] of this.dependencies) {
          if (to === id) {
            targetGraph.push([from, to]);

            if (from) {
              queue.push(from);
            }
          }
        }
      }
    }

    return targetGraph;
  }

  loadConfig(config: Config) {
    this.config = config;

    for (const [id, targetDefinition] of Object.entries(this.config.pipeline)) {
      this.addTargetDefinition(id, targetDefinition);
    }

    // add target definitions for unknown tasks
    const knownTasks = new Set<string>();
    for (const target of this.targets.values()) {
      knownTasks.add(target.task);
    }

    knownTasks.add(START_TARGET_ID);

    const unknownCommands = this.config.command.filter((cmd) => !knownTasks.has(cmd));

    for (const command of unknownCommands) {
      this.addTargetDefinition(command, [`^${command}`]);
    }

    this.addDependencies();
  }

  private getTargetPriority(target: PipelineTarget) {
    return target.priority !== undefined
      ? target.priority
      : this.config.priorities?.find((priority) => priority.package === target.packageName && priority.task === target.task)?.priority;
  }

  /**
   * The "run" public API, accounts for setting distributed mode for the master lage node
   *
   * Runs the pipeline with the p-graph library
   *
   * Note: this is the abstraction layer on top of the external p-graph library to insulate
   *       any incoming changes to the library.
   */
  async run(context: RunContext) {
    this.context = context;

    const nodeMap: PGraphNodeMap = new Map();
    const targetGraph = this.generateTargetGraph();

    for (const [from, to] of targetGraph) {
      const fromTarget = this.targets.get(from)!;
      const toTarget = this.targets.get(to)!;

      for (const target of [fromTarget, toTarget]) {
        nodeMap.set(target.id, {
          run: () => {
            if (target.id === START_TARGET_ID || !target.run) {
              return Promise.resolve();
            }

            const wrappedTask = new WrappedTarget(target, this.workspace.root, this.config, context);
            return wrappedTask.run();
          },
          priority: this.getTargetPriority(target),
        });
      }
    }

    await pGraph(nodeMap, targetGraph).run({
      concurrency: this.config.concurrency,
      continue: this.config.continue,
    });
  }
}
