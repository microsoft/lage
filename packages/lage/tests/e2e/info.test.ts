import { Monorepo } from "../mock/monorepo";
import { getTargetId } from "../../src/task/taskId";
import { parseNdJson } from "./parseNdJson";
import { silent } from "npmlog";
import exp from "constants";

describe("info test", () => {
  it("with json reporter", () => {
    const result = testInfoWithReporter("json");
    var resultObj = JSON.parse(result);
    // make deterministic
    resultObj.timestamp = 0;
    expect(JSON.stringify(resultObj)).toMatchSnapshot();
  });

  it("with dgml reporter", () => {
    const result = testInfoWithReporter("dgml");
    expect(result).toMatchSnapshot();
  });
});

function testInfoWithReporter(reporterName: string): string {
  const repo = new Monorepo("info-" + reporterName);

  repo.init();
  repo.install();

  repo.addPackage("FooApp1", ["FooCore"]);
  repo.addPackage("FooApp2", ["FooCore"]);
  repo.addPackage("FooCore", ["BuildTool"]);
  repo.addPackage("BarPage", ["BarCore"]);
  repo.addPackage("BarCore", ["BuildTool"]);
  repo.addPackage("BuildTool");
  repo.linkPackages();

  const results = repo.run("writeInfo", ["test", "--reporter", reporterName], true);
  expect(results.exitCode).toBe(0);
  expect(results.stderr).toBe("");

  repo.cleanup();

  return results.stdout;
}
