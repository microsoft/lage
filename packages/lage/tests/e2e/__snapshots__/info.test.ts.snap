// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`info test with dgml reporter 1`] = `"<DirectedGraph GraphDirection="LeftToRight" xmlns="http://schemas.microsoft.com/vs/2009/dgml"><Nodes><Node Id="BarCore#test" Label="BarCore#test"><Category Ref="test"/></Node><Node Id="BarPage#test" Label="BarPage#test"><Category Ref="test"/></Node><Node Id="BuildTool#test" Label="BuildTool#test"><Category Ref="test"/></Node><Node Id="FooApp1#test" Label="FooApp1#test"><Category Ref="test"/></Node><Node Id="FooApp2#test" Label="FooApp2#test"><Category Ref="test"/></Node><Node Id="FooCore#test" Label="FooCore#test"><Category Ref="test"/></Node><Node Id="BarCore#build" Label="BarCore#build"><Category Ref="build"/></Node><Node Id="BarPage#build" Label="BarPage#build"><Category Ref="build"/></Node><Node Id="BuildTool#build" Label="BuildTool#build"><Category Ref="build"/></Node><Node Id="FooApp1#build" Label="FooApp1#build"><Category Ref="build"/></Node><Node Id="FooApp2#build" Label="FooApp2#build"><Category Ref="build"/></Node><Node Id="FooCore#build" Label="FooCore#build"><Category Ref="build"/></Node></Nodes><Links><Link Source="BarCore#test" Target="BarCore#build"/><Link Source="BarPage#test" Target="BarPage#build"/><Link Source="BuildTool#test" Target="BuildTool#build"/><Link Source="FooApp1#test" Target="FooApp1#build"/><Link Source="FooApp2#test" Target="FooApp2#build"/><Link Source="FooCore#test" Target="FooCore#build"/><Link Source="BarCore#build" Target="BuildTool#build"/><Link Source="BarPage#build" Target="BarCore#build"/><Link Source="FooApp1#build" Target="FooCore#build"/><Link Source="FooApp2#build" Target="FooCore#build"/><Link Source="FooCore#build" Target="BuildTool#build"/></Links><Categories><Category Id="test" Label="test"/><Category Id="build" Label="build"/></Categories></DirectedGraph>"`;

exports[`info test with json reporter 1`] = `"{"timestamp":0,"level":30,"msg":"info","data":{"command":["test"],"scope":["BarCore","BarPage","BuildTool","FooApp1","FooApp2","FooCore"],"packageTasks":[{"id":"BarCore#test","command":["npm","run","test"],"dependencies":["BarCore#build"],"workingDirectory":"packages/BarCore","package":"BarCore","task":"test"},{"id":"BarPage#test","command":["npm","run","test"],"dependencies":["BarPage#build"],"workingDirectory":"packages/BarPage","package":"BarPage","task":"test"},{"id":"BuildTool#test","command":["npm","run","test"],"dependencies":["BuildTool#build"],"workingDirectory":"packages/BuildTool","package":"BuildTool","task":"test"},{"id":"FooApp1#test","command":["npm","run","test"],"dependencies":["FooApp1#build"],"workingDirectory":"packages/FooApp1","package":"FooApp1","task":"test"},{"id":"FooApp2#test","command":["npm","run","test"],"dependencies":["FooApp2#build"],"workingDirectory":"packages/FooApp2","package":"FooApp2","task":"test"},{"id":"FooCore#test","command":["npm","run","test"],"dependencies":["FooCore#build"],"workingDirectory":"packages/FooCore","package":"FooCore","task":"test"},{"id":"BarCore#build","command":["npm","run","build"],"dependencies":["BuildTool#build"],"workingDirectory":"packages/BarCore","package":"BarCore","task":"build"},{"id":"BarPage#build","command":["npm","run","build"],"dependencies":["BarCore#build"],"workingDirectory":"packages/BarPage","package":"BarPage","task":"build"},{"id":"BuildTool#build","command":["npm","run","build"],"dependencies":[],"workingDirectory":"packages/BuildTool","package":"BuildTool","task":"build"},{"id":"FooApp1#build","command":["npm","run","build"],"dependencies":["FooCore#build"],"workingDirectory":"packages/FooApp1","package":"FooApp1","task":"build"},{"id":"FooApp2#build","command":["npm","run","build"],"dependencies":["FooCore#build"],"workingDirectory":"packages/FooApp2","package":"FooApp2","task":"build"},{"id":"FooCore#build","command":["npm","run","build"],"dependencies":["BuildTool#build"],"workingDirectory":"packages/FooCore","package":"FooCore","task":"build"}]}}"`;
